packages/editor/src/hooks/block-bindings.js

```diff
@@ -0,0 +1,303 @@
+	);
+}
+
+function BlockBindingsAttribute( {
+	toggleProps,
+	attribute,
+	filteredBindings,
+} ) {
+	return (
+		<ItemGroup>
+			<Button { ...toggleProps }>
+				<Hstack align="center" justify="flex-start" expanded={ false }>
+					<Icon icon={ customPostType } />
+					<Truncate numberOfLines={ 1 } ellipsis="…">
+						{ attribute }
+					</Truncate>
+					{ !! filteredBindings[ attribute ] && (
+						<>
+							<Icon icon={ chevronRightSmall } />
+							<Truncate
+								numberOfLines={ 1 }
+								ellipsis="…"
+								className="components-panel__block-bindings-panel-item-source"
+							>
+								{ filteredBindings[ attribute ]?.args?.key }
+							</Truncate>
+						</>
+					) }
+				</Hstack>
+			</Button>
+		</ItemGroup>
+	);
+}
+
+const BlockBindingsPanel = ( { name, attributes: { metadata } } ) => {
+	const { bindings } = metadata || {};
+
+	const bindableAttributes = getBindableAttributes( name );
+	const dropdownMenuProps = useToolsPanelDropdownMenuProps();
+
+	const filteredBindings = { ...bindings };
+	Object.keys( filteredBindings ).forEach( ( key ) => {
+		if (
+			! canBindAttribute( name, key ) ||
+			filteredBindings[ key ].source === 'core/pattern-overrides'
+		) {
+			delete filteredBindings[ key ];
+		}
+	} );
+
+	const { updateBlockAttributes } = useDispatch( blockEditorStore );
+
+	const { _id } = useSelect( ( select ) => {
+		const { getSelectedBlockClientId } = select( blockEditorStore );
+
+		const selectedBlockClientId = getSelectedBlockClientId();
+		return {
+			_id: selectedBlockClientId,
+		};
+	}, [] );
+
+	const removeAllConnections = () => {
+		const newMetadata = { ...metadata };
+		delete newMetadata.bindings;
+		updateBlockAttributes( _id, {
+			metadata:
+				Object.keys( newMetadata ).length === 0
+					? undefined
+					: newMetadata,
+		} );
+	};
+
+	const addConnection = ( value, attribute ) => {
+		// Assuming the block expects a flat structure for its metadata attribute
+		const newMetadata = {
+			...metadata,
+			// Adjust this according to the actual structure expected by your block
+			bindings: {
+				...metadata?.bindings,
+				[ attribute ]: {
+					source: 'core/post-meta',
+					args: { key: value },
+				},
+			},
+		};
+		// Update the block's attributes with the new metadata
+		updateBlockAttributes( _id, {
+			metadata: newMetadata,
+		} );
+	};
+
+	const removeConnection = ( key ) => {
+		const newMetadata = { ...metadata };
+		if ( ! newMetadata.bindings ) {
+			return;
+		}
+
+		delete newMetadata.bindings[ key ];
+		if ( Object.keys( newMetadata.bindings ).length === 0 ) {
+			delete newMetadata.bindings;
+		}
+		updateBlockAttributes( _id, {
+			metadata:
+				Object.keys( newMetadata ).length === 0
+					? undefined
+					: newMetadata,
+		} );
+	};
+	const postMeta = useSelect( ( select ) => {
+		return select( editorStore ).getEditedPostAttribute( 'meta' );
+	}, [] );
+	if ( postMeta === undefined || bindableAttributes.length === 0 ) {
+		return null;
+	}
```

Comment:
Taking a closer look at the code now, I believe it is worth trying moving this to the source definition and just read the sources from here. Basically, I'm suggesting something like this:

```js
	const fieldsList = {};
	Object.values( registeredSources ).forEach(
		( { getFieldsList, label } ) => {
			if ( getFieldsList ) {
				fieldsList[ label ] = getFieldsList( {
					registry,
					context,
				} );
			}
		}
	);

	if ( Object.keys( fieldsList ).length === 0 ) {
		return null;
	}
```

And then, define the `getFieldsList` in each source. For example, in [post meta](https://github.com/WordPress/gutenberg/blob/9a5fcaf9e3dfe43fc3f96a4227c5eac054d9e9a4/packages/editor/src/bindings/post-meta.js), we would add a new property like this:

```js
	getFieldsList( { registry, context } ) {
                 // TODO: Filter footnotes.
		return registry
			.select( coreDataStore )
			.getEditedEntityRecord(
				'postType',
				context?.postType,
				context?.postId
			).meta;
	},
```
This way:
* We could avoid moving the hooks to the editor package. This would be the main reason.
* The API is still private so we can iterate on that.
* It'd be a first step to allow other sources to hook into this list. If we want another source to appear there, they would just need to add that property to the source registration.

Let me know if what you think 🙂

Reply:
That way you are actually allowing custom sources to list their own fields. Being post meta one of them. Pretty smart!